portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java

```diff
@@ -0,0 +1,62 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import java.io.IOException;
+import java.util.List;
+import java.util.ArrayList;
+import com.google.gson.Gson;
+import com.google.sps.classes.Comment;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+@WebServlet("/comments")
+public class CommentsServlet extends HttpServlet {
+    
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.setContentType("application/json;");
+
+        Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery comments = datastore.prepare(query);
+
+        List<Comment> comments_to_set = new ArrayList<>();
+        for (Entity entity : comments.asIterable()) {
+            String username = (String) entity.getProperty("username");
+            String comment = (String) entity.getProperty("comment");
+            long timestamp = (long) entity.getProperty("timestamp");
+
+            Comment comment_to_set = new Comment(username, comment, timestamp);
+            comments_to_set.add(comment_to_set);
+        }
+
+        List<Comment> newest_comments = comments_to_set.subList(0, comments_to_set.size() > 5 ? 5 : comments_to_set.size());
```

Comment:
Maybe just cut off the previous loop after 5 iterations? That way you can save a couple of passes over the ArrayLists.

You could mess around with the for loop statement itself or just keep a separate counter and at the top of the for-loop break out after you hit 5. Something like:

```java
int i = 0;
int num_comments = 5;
for ... {
    if (i >= num_comments) {
        break;
    }
...
    i++;
}
```

I'd actually recommend changing the for-loop statement to break on either the end of comments.asIterable or a counter hitting 5, but I don't know exactly how that'd look :)

Reply:
I actually ended removing this logic since the walk through requested a select option for the user to limit the amount of comments to fetch. But yeah, I believe your approach with the for loop is better! 😄